src/v/archival/ntp_archiver_service.cc

```diff
@@ -854,4 +871,137 @@ ntp_archiver::maybe_truncate_manifest(retry_chain_node& rtc) {
     co_return result;
 }
 
+ss::future<> ntp_archiver::housekeeping() {
+    try {
+        if (housekeeping_can_continue()) {
+            co_await apply_retention();
+            co_await garbage_collect();
+        }
+    } catch (std::exception& e) {
+        vlog(_rtclog.warn, "Error occured during housekeeping", e.what());
+    }
+}
+
+ss::future<> ntp_archiver::apply_retention() {
+    if (!housekeeping_can_continue()) {
+        co_return;
+    }
+
+    auto retention_calculator = retention_calculator::factory(
+      manifest(), _partition->get_ntp_config());
+    if (!retention_calculator) {
+        co_return;
+    }
+
+    auto next_start_offset = retention_calculator->next_start_offset();
+    if (next_start_offset) {
+        vlog(
+          _rtclog.debug,
+          "{} Advancing start offset to {} satisfy retention policy",
+          retention_calculator->strategy_name(),
+          *next_start_offset);
+
+        auto sync_timeout = config::shard_local_cfg()
+                              .cloud_storage_metadata_sync_timeout_ms.value();
+        auto deadline = ss::lowres_clock::now() + sync_timeout;
+        auto error = co_await _partition->archival_meta_stm()->truncate(
+          *next_start_offset, deadline, _as);
+        if (error != cluster::errc::success) {
+            vlog(
+              _rtclog.warn,
+              "Failed to update archival metadata STM start offest according "
+              "to retention policy: {}",
+              error);
+        }
+    } else {
+        vlog(
+          _rtclog.debug,
+          "{} Retention policies are already met.",
+          retention_calculator->strategy_name());
+    }
+}
+
+// Garbage collection can be improved as follows:
+// * issue #6843: delete via DeleteObjects S3 api instead of deleting individual
+// segments
+// * issue #6844: flush _replaced and advance the start offset even if the
+// deletions fail if the backlog breaches a certain limit. This can lead to
+// segments begin orphaned, but it's preferable to unbounded backlog growth.
+ss::future<> ntp_archiver::garbage_collect() {
+    if (!housekeeping_can_continue()) {
+        co_return;
+    }
+
+    auto to_remove = _partition->archival_meta_stm()->get_segments_to_cleanup();
+
+    std::atomic<size_t> successful_deletes{0};
```

Comment:
@VladLazar i don't see any reason you would need an atomic here. is there something i'm not seeing? if not, can you change this to a plain `size_t` since std::atomic will blow cache lines...

Reply:
`succesful_deletes` can be incremented concurrently by the futures spawned by `ss::max_concurrent_for_each` below. I could probably rework this code to return a list of futures and infer how many deletions where successful from that.

Reply:
> succesful_deletes can be incremented concurrently by the futures spawned by ss::max_concurrent_for_each below. I could probably rework this code to return a list of futures and infer how many deletions where successful from that.

how so? even if multiple concurrent continuations are runnable, only one at a time will run on the same core. are you saying there is physical concurrency across cores here?

Reply:
You're right. Thanks for pointing it out. There's no physical concurrency here. The word "parallel" in the docstring made me think that `max_concurrent_for_each` goes across cores. Looking at the code, it doesn't.